File: src/plugin/menus/TextMenu.java
package plugin.menus;

import arc.Events;
import arc.func.Cons2;
import arc.struct.ObjectMap;
import arc.util.Timer;
import mindustry.game.EventType;
import mindustry.gen.Call;
import mindustry.gen.Player;

public class TextMenu {
    static int lastId = 0;
    static ObjectMap<Integer, TextMenu> menus = new ObjectMap<>();

    Player player;
    int id;
    Cons2<Player, String> handler;
    String title = "No title", message = "Write here", defMessage = "";
    int len = 32;
    boolean numeric = false;
    Timer.Task invalidTimer = null;

    TextMenu(Cons2<Player, String> handler) {
        this.handler = handler;
        lastId += 1;
        this.id = lastId;
    }

    public TextMenu setTitle(String var) {
        this.title = var;
        return this;
    }

    public TextMenu setMessage(String var) {
        this.message = var;
        return this;
    }

    public TextMenu setDefMessage(String var) {
        this.defMessage = var;
        return this;
    }

    public TextMenu setLen(int var) {
        this.len = var;
        return this;
    }

    public TextMenu setNumeric(boolean var) {
        this.numeric = var;
        return this;
    }

    public TextMenu show(Player p) {
        Call.textInput(p.con, id, title, message, len, defMessage, numeric);

        this.player = p;

        menus.put(id, this);

        invalidTimer = Timer.schedule(this::cancel, 60f);

        return this;
    }

    void cancel() {
        menus.remove(id);
        if(invalidTimer != null) {
            invalidTimer.cancel();
            invalidTimer = null;
        }
    }

    public static void load() {
        Events.on(EventType.PlayerLeave.class, e -> {
            menus.each((id, menu) -> {
                if (menu.player == e.player) menu.cancel();
            });
        });

        Events.on(EventType.TextInputEvent.class, (e) -> {
            Player player = e.player;
            String text = e.text;

            TextMenu menu = menus.get(e.textInputId);
            if (menu == null || menu.player != player) {
                player.sendMessage("[scarlet]Unknown input!");
                return;
            }

            menu.cancel();
            menu.handler.get(player, text);
        });
    }
}

File: src/main/kotlin/plugin/menus/TextMenu.kt
// stolen from old escoplugin + fixed by ai
package plugin.menus

import arc.Events
import arc.func.Cons2
import arc.struct.ObjectMap
import arc.struct.Seq
import arc.util.Timer
import mindustry.game.EventType
import mindustry.gen.Call
import mindustry.gen.Player
import plugin.Bundle

class TextMenu(
    private val handler: Cons2<Player, String>,
) {
    lateinit var player: Player
    val id: Int = ++lastId

    var title: String = "No title"
    var message: String = "Write here"
    var defMessage: String = ""
    var len: Int = 32
    var numeric: Boolean = false

    private var invalidTimer: Timer.Task? = null

    fun setTitle(value: String): TextMenu {
        title = value
        return this
    }

    fun setMessage(value: String): TextMenu {
        message = value
        return this
    }

    fun setDefMessage(value: String): TextMenu {
        defMessage = value
        return this
    }

    fun setLen(value: Int): TextMenu {
        len = value
        return this
    }

    fun setNumeric(value: Boolean): TextMenu {
        numeric = value
        return this
    }

    fun show(p: Player): TextMenu {
        Call.textInput(p.con, id, title, message, len, defMessage, numeric)

        player = p

        menus.put(id, this)

        invalidTimer = Timer.schedule({ cancel() }, 60f)

        return this
    }

    fun cancel() {
        menus.remove(id)
        invalidTimer?.cancel()
        invalidTimer = null
    }

    companion object {
        var lastId = 0
        val menus = ObjectMap<Int, TextMenu>()

        fun load() {
            Events.on(EventType.PlayerLeave::class.java) { e ->
                val toRemove = Seq<Int>()
                menus.each { id, menu ->
                    if (menu.player == e.player) toRemove.add(id)
                }
                toRemove.each { id ->
                    menus.get(id)?.cancel()
                }
            }

            Events.on(EventType.TextInputEvent::class.java) { e ->
                val player = e.player
                val text = e.text

                val menu = menus.get(e.textInputId)
                if (menu == null || menu.player != player) {
                    player.sendMessage(Bundle.get("menu.invalid-input", player.locale))
                    return@on
                }

                menu.cancel()
                menu.handler.get(player, text)
            }
        }
    }
}
